Show "Create another" checkbox on new resource page.

bolshakov opened the pull request:
If this checkbox is selected, user will be redirected to new resource page after successfully creating current record.
This feature may be very helpful when entering large amount of data and it's nice to have such feature in AA itself.

Currently build on master is broken, so I failed to run rails server and verify it works. But, it should looks this way: 

![screenshot 2016-06-18 12 08 17](https://cloud.githubusercontent.com/assets/991265/16186507/f317b0c2-36db-11e6-9e1d-146d99b353af.png)

I'm not happy with implementation and glad to receive any feedback to improve my solution.

bolshakov commented on Gemfile:
```diff
@@ -11,6 +11,8 @@ gem 'jquery-ui-rails', rails_version[0] == '3' ? '~> 4.0' : '~> 5.0'
 
 gem 'test-unit', '~> 3.0' if rails_version[0] == '3'
 
+gem 'arbre', github: 'activeadmin/arbre'
```
[This change](https://github.com/activeadmin/arbre/pull/49) required, but new arbre version is not published

timoschilling replied:
This is no longer necessary, the change is released in `1.1.1`, but Arbre `1.1.1` should be the new min version in `activeadmin.gemfile`.

bolshakov replied:
@timoschilling I updated minimum arbre version and rebased on top of master. Sorry for delay

bolshakov commented on features/step_definitions/web_steps.rb:
```diff
@@ -72,7 +72,12 @@ def with_scope(locator)
 
 Then /^the "([^"]*)" checkbox(?: within (.*))? should( not)? be checked$/ do |label, parent, negate|
   with_scope(parent) do
-    expect(find_field(label)['checked']).to negate ? eq(false) : eq(true)
```
previous solution does not work with the following checked checkbox:

```html
<input checked="on" ...>
```

bolshakov commented on lib/active_admin/resource_controller/data_access.rb:
```diff
@@ -312,6 +312,24 @@ def assign_attributes(resource, attributes)
+      # @return [String]
+      def smart_resource_url
+        if create_another?
+          # FIXME: on rails 5, it should be "permitted_params.to_h.slice(:create_another)"
```
Should I write code for all rails versions?

bolshakov commented on lib/active_admin/views/components/active_admin_form.rb:
```diff
@@ -72,10 +72,24 @@ def actions(*args, &block)
+              input(create_another, checked: create_another, type: 'checkbox', id: 'create_another', name: 'create_another')
+              label(I18n.t('active_admin.create_another'), for: 'create_another')
+            end
+          end
```
This hack is required since `ActiveAdminForm` delegates `input` method to form builder. Thus, I have to create new arbre context to bypass this override. Would you like to see this comment in the code?

Is there a simpler way to render checkbox here?

faucct commented:
Does the checkbox stay checked after the redirect? If not then you do the same amount of actions as when clicking 'New Resource' and the only boost of speed you get is from skipping the 'show' page render.

bolshakov commented:
@faucct yes, this checkbox is checked after save https://github.com/activeadmin/activeadmin/pull/4477/files#diff-9868506c30b8ad5139d31b77f6099bc8R41

timoschilling commented:
I like this idea!

timoschilling commented on lib/active_admin/resource_controller/data_access.rb:
```diff
@@ -312,6 +312,24 @@ def assign_attributes(resource, attributes)
+      # @return [Boolean] true if user requested to create one more
+      #   resource after creating this one.
+      def create_another?
+        permitted_params[:create_another] == "on"
```
I think we only need to change the presence of the param and not the value.

timoschilling commented on config/locales/en.yml:
```diff
@@ -8,6 +8,7 @@ en:
     edit: "Edit"
     delete: "Delete"
     delete_confirmation: "Are you sure you want to delete this?"
+    create_another: "Create another"
```
Should be `Create another %{model}`

timoschilling commented:
Currently I try to use this PR in one of my private projects and I get a CSS issue:

timoschilling commented:
We should have a application, namespace and resource level configuration for that.

timoschilling commented:
My project use Rails 5 but not `permitted_params`, while it is only a tool for me.
And I get this error on update and create:
<img width="612" alt="action_controller__exception_caught" src="https://cloud.githubusercontent.com/assets/165599/17565042/fdae302e-5f35-11e6-8995-505a84e30b81.png">

timoschilling commented:
If I checked the checkbox, it should stay checked.

bolshakov commented:
@timoschilling I fixed most of the issues except configuration and CSS.

> Currently I try to use this PR in one of my private projects and I get a CSS issue: 

Have you tasted with the latest PR version? I think it should work now.

> We should have a application, namespace and resource level configuration for that.

Could you refer for some example of such configuration for reference?

bolshakov commented:
@timoschilling I added configuration, fixed styling issue and fixed tests on rails 5. Some builds still failing because of dependency issues. I'm not sure if this is mine fault.

bolshakov commented:
@timoschilling any chances to merge?

varyonic commented:
@bolshakov I don't have [edit access to this PR](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) so please rebase and we will review.

bolshakov commented:
@varyonic I've rebased and enabled edit access

Fivell commented on lib/active_admin/application.rb:
```diff
@@ -112,7 +116,7 @@ def initialize
     # Request parameters that are permitted by default
     inheritable_setting :permitted_params, [
-      :utf8, :_method, :authenticity_token, :commit, :id
+      :utf8, :_method, :authenticity_token, :commit, :id, :create_another
```
```: create_another``` should be part of permitted_params if only enabled

Fivell commented on lib/active_admin/resource_controller/data_access.rb:
```diff
@@ -298,6 +298,29 @@ def assign_attributes(resource, attributes)
+      def smart_resource_url
+        if create_another?
+          create_another_checkbox =
+            if Rails::VERSION::MAJOR >= 4
```
looks like this checking ```Rails::VERSION::MAJOR``` no needed anymore

Fivell commented on lib/active_admin/resource_dsl.rb:
```diff
@@ -66,8 +66,12 @@ def permit_params(*args, &block)
+          permitted_params =
+            active_admin_namespace.permitted_params +
+              Array.wrap(belongs_to_param) +
+              Array.wrap((:create_another if active_admin_namespace.create_another?))
```
why checking ```namespace```? IMO it should check create_another on resource.

Fivell commented:
@bolshakov , the last thing is changelog entry and few lines in docs about this required

bolshakov commented:
@Fivell done

Fivell commented:
@bolshakov ,  I think ```active_admin_namespace.create_another?``` should be changed to ```config.create_another```  in order to make tests passing

Fivell commented:
@bolshakov, 👍  for contribution, but there is one more change required: can you please make it disabled by default, thanks

bolshakov commented:
@Fivell sure!

Fivell commented:
@bolshakov , please update documentation about default behaviour, also take a look at failing specs

bolshakov commented:
@Fivell sorry for the noise. Specs and docs fixed

Fivell commented on docs/5-forms.md:
```diff
@@ -173,3 +173,23 @@ You can arrange content in tabs as shown below:
+
+## Customize the Create Another checkbox
+
+By default ActiveAdmin shows "Create another Resource" checkbox on a new resource page. You may disable this 
```
@bolshakov , By default ActiveAdmin doesn't show "Create another Resource"

Fivell commented on lib/generators/active_admin/install/templates/active_admin.rb.erb:
```diff
@@ -179,6 +179,13 @@ ActiveAdmin.setup do |config|
 
+  # == Removing Create Another checkbox
+  # 
+  # Create another checkbox is enabled by default. You can customize it for individual
```
disabled by default

bolshakov commented:
Woohoo!!!
